### PR TITLE
feat(nfts): trailblazers badges s2

### PR DESCRIPTION
**Pull request by bearni95**
- [x] s2 migration with open/close windows
- [x] deployment scripts
- [x] tests
- [x] determine max tampers via level

### CODE REVIEW

File: packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        uint256 leagueId;
+        uint256 color; // 0 = neutral, 1 = pink, 2 = purple
+        uint256 power;
+    }
+
+    struct League {
+        uint256 openTime; // registration starts
+        uint256 closeTime; // registration ends
+        uint256 startTime; // league starts (requires admin action)
+        uint256 seed;
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol:40]: use uint64 to make League use only one slot?

File: packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        uint256 _leagueId,
+        address _owner,
+        address _badgeContract,
+        uint256 _tokenId
+    )
+        public
+        pure
+        returns (bytes memory)
+    {
+        return abi.encodePacked(_owner, _badgeContract, _tokenId, _leagueId);
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol:70]: why not hash it to bytes32?

File: packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        view
+        returns (uint256 openTime, uint256 closeTime, uint256 startTime, uint256 seed)
+    {
+        return getLeague(currentLeagueId);
+    }
+
+    function getLeague(uint256 _leagueId)
+        public
+        view
+        returns (uint256 openTime, uint256 closeTime, uint256 startTime, uint256 seed)
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol:117]: why not return (League memory league)?

File: packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        League memory league =
+            League({ openTime: _openTime, closeTime: _closeTime, startTime: _startTime, seed: 0 });
+        currentLeagueId += 1;
+
+        leagues[currentLeagueId] = league;
+
+        emit LeagueCreated(currentLeagueId, _openTime, _closeTime, _startTime);
+    }
+
+    modifier leagueOpen(uint256 _leagueId) {
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-badges/BadgeChampions.sol:140]: I think all modifiers shall go before functions, can you lint your code and fix all issues?

**Review by dantaik — CHANGES_REQUESTED** [review]

**Comment by dantaik** [comment]
I don't know enough about the overall design, so probably ask someone else to review the logics.

**Review by KorbinianK — APPROVED** [review]

File: packages/nfts/contracts/trailblazers-season-2/BadgeMigration.sol [hunk 5]
```diff
@@ -0,0 +1,606 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/extensions/ERC1155SupplyUpgradeable.sol";
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-season-2/BadgeMigration.sol:4]: please add new contracts into [gen-layouts.sh](https://github.com/taikoxyz/taiko-mono/blob/main/packages/nfts/deployments/gen-layouts.sh) so we always update the [layout file](https://github.com/taikoxyz/taiko-mono/blob/main/packages/nfts/contract_layout.md) for comparison.

One example: https://github.com/taikoxyz/taiko-mono/pull/18276/files

File: packages/nfts/contracts/trailblazers-season-2/TrailblazersBadgesS2.sol [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+        minter = _minter;
+        uriTemplate = _uriTemplate;
+    }
+
+    /// @notice Set the minter address
+    /// @param _minter The minter address
+    /// @dev Only the owner can call this function
+    function setMinter(address _minter) external virtual onlyOwner {
+        minter = _minter;
```
Comment (dantaik) [inline packages/nfts/contracts/trailblazers-season-2/TrailblazersBadgesS2.sol:105]: why not use `onlyRole`

**Review by dantaik — APPROVED** [review]

**Review by bennettyong — APPROVED** [review]

**Review by dantaik — APPROVED** [review]

File: .github/workflows/guardian-prover-health-check-ui.yml [hunk 7]
```diff
@@ -28,8 +28,6 @@ jobs:
   #     vercel_org_id: ${{ secrets.VERCEL_ORG_ID }}
```
Comment (KorbinianK) [inline .github/workflows/guardian-prover-health-check-ui.yml:28]: this PR shouldn't touch other packages

File: packages/docs-site/src/content/docs/core-concepts/contestable-rollups.md [hunk 8]
```diff
@@ -23,5 +23,4 @@ If the original prover wins: The original prover reclaims their validity bond an
```
Comment (KorbinianK) [inline packages/docs-site/src/content/docs/core-concepts/contestable-rollups.md:23]: this PR shouldn't touch other packages

File: packages/docs-site/src/content/docs/network-reference/software-releases-and-deployments.md [hunk 9]
```diff
@@ -9,11 +9,11 @@ It is **highly recommended** you use the latest software. You can find the lates
 
 | Package                                                                                | Release notes                                                                                                                                                                                     |
```
Comment (KorbinianK) [inline packages/docs-site/src/content/docs/network-reference/software-releases-and-deployments.md:10]: this PR shouldn't touch other packages

File: packages/protocol/contract_layout_layer1.md
Comment (KorbinianK) [inline packages/protocol/contract_layout_layer1.md:1]: this PR shouldn't touch other packages

File: packages/protocol/contract_layout_layer2.md
Comment (KorbinianK) [inline packages/protocol/contract_layout_layer2.md:1]: this PR shouldn't touch other packages

**Review by KorbinianK — APPROVED** [review]

**Pull request merged**